src/metabase/related.clj

```diff
@@ -0,0 +1,274 @@
+(defn- linked-from
+  [table]
+  (let [fields (db/select-field :id Field :table_id (:id table))]
+    (->> (db/select-field :table_id Field
+           :fk_target_field_id [:in fields])
+         (map Table)
+         (filter (every-pred (comp nil? :visibility_type) mi/can-read?))
+         (take max-matches))))
+
+(defn- cards-sharing-dashboard
+  [card]
+  (when-let [dashboards (not-empty (db/select-field :dashboard_id DashboardCard
+                                     :card_id (:id card)))]
+    (->> (db/select-field :card_id DashboardCard
+           :dashboard_id [:in dashboards]
+           :card_id [:not= (:id card)])
+         (map Card)
+         (filter (every-pred (comp false? :archived) mi/can-read?))
+         (take max-matches))))
+
+(defn- similar-questions
+  [card]
+  (->> (db/select Card
+         :table_id (:table_id card)
+         :archived false)
+       (filter mi/can-read?)
+       (rank-by-similarity card)
+       (filter (comp pos? :similarity))))
+
+(defn- canonical-metric
+  [card]
+  (->> (db/select Metric
+         :table_id (:table_id card)
+         :is_active true)
+       (filter (every-pred mi/can-read?
+                           (comp #{(-> card
+                                       :dataset_query
+                                       :query
+                                       :aggregation)}
+                                 :aggregation :definition)))
+       first))
+
+(defn- recently-modified-dashboards
+  []
+  (->> (db/select-field :model_id 'Revision
+         :model   "Dashboard"
+         :user_id api/*current-user-id*
+         {:order-by [[:timestamp :desc]]})
+       (map Dashboard)
+       (filter (every-pred (comp false? :archived) mi/can-read?))
+       (take max-serendipity-matches)))
+
+(defn- recommended-dashboards
+  [cards]
+  (let [recent           (recently-modified-dashboards)
+        card->dashboards (->> (apply db/select [DashboardCard :card_id :dashboard_id]
+                                     (cond-> {}
+                                       (not-empty cards)
+                                       (assoc :card_id [:in (map :id cards)])
+
+                                       (not-empty recent)
+                                       (assoc :dashboard_id [:not-in recent])))
+                              (group-by :card_id))
+        best             (->> cards
+                              (mapcat (comp card->dashboards :id))
+                              distinct
+                              (map Dashboard)
+                              (filter (every-pred (comp false? :archived)
+                                                  mi/can-read?))
+                              (take max-best-matches))]
+    (concat best recent)))
+
+(defn- recommended-collections
+  [cards]
+  (->> cards
+       (m/distinct-by :collection_id)
+       interesting-mix
+       (keep (comp Collection :collection_id))
+       (filter (every-pred (comp false? :archived) mi/can-read?))))
```

Comment:
I'm not a huge fan of how every single recommender is doing some form of `my/can-read?`. Seems like something that should be abstracted away imo